File: code/reporter/src/app/db/crud/target.py
from datetime import datetime
from bson.objectid import ObjectId

from app.db.models.target import Target
from app.db.crud.base import CrudBase
from app.enums.main import ScanStatus


class CrudTarget(CrudBase):
    COLLECTION_NAME = "targets"

    def __init__(self, db_session) -> None:
        super().__init__(db_session, self.COLLECTION_NAME)

    def find_by_id(self, target_id: ObjectId) -> Target | None:
        """
        Find Target by id

        Args:
            target_id (ObjectId): _description_

        Raises:
            Exception: _description_

        Returns:
            Target | None: _description_
        """
        if not target_id or not isinstance(target_id, ObjectId):
            raise Exception("Invalid Mongodb object id given")

        row = self.collection.find_one({"_id": target_id})
        if not row:
            return None

        # Change _id to id
        row.update({"id": row["_id"]})
        return Target.model_validate(row)

    def update_scan_status(self, target: Target):
        """
        Update Scan Status by target

        Args:
            target (Target): _description_

        Raises:
            Exception: _description_
        """
        if not target or not isinstance(target, Target):
            raise Exception("Invalid Target object given")

        now = datetime.now()
        to_update = {}
        if target.scan_status == ScanStatus.ENUM.SCAN_STARTED:
            target.scan_started_time = now
            to_update["scan_started_time"] = now
        elif target.scan_status == ScanStatus.ENUM.REPORT_GENERATED:
            target.scan_completed_time = now
            to_update["scan_completed_time"] = now

        to_update["scan_status"] = target.scan_status
        self.collection.update_one({"_id": target.id}, {"$set": to_update})

    def update_scan_status_by_id(self, target_id: ObjectId, scan_status):
        if scan_status is None:
            raise Exception("Invalid scan_status")

        if not target_id or not isinstance(target_id, ObjectId):
            raise Exception("Invalid Mongodb object id given")

        now = datetime.now()

        to_update = {}
        if scan_status == ScanStatus.ENUM.SCAN_STARTED:
            to_update["scan_started_time"] = now
        elif scan_status == ScanStatus.ENUM.REPORT_GENERATED:
            to_update["scan_completed_time"] = now

        to_update["scan_status"] = scan_status

        self.collection.update_one({"_id": target_id}, {"$set": to_update})

    def mark_as_completed(self, target: Target):
        """
        Update Scan Status by target

        Args:
            target (Target): _description_

        Raises:
            Exception: _description_
        """
        if not target or not isinstance(target, Target):
            raise Exception("Invalid Target object given")

        now = datetime.now()
        target.scan_status = ScanStatus.ENUM.REPORT_GENERATED
        target.scan_completed_time = now

        to_update = {
            "scan_completed_time": now,
            "scan_status": target.scan_status,
            "overall_cvss_score": target.overall_cvss_score,
        }

        if target.is_ip_range():
            to_update["cvss_score_by_host"] = target.cvss_score_by_host

        to_update["scan_status"] = target.scan_status
        self.collection.update_one({"_id": target.id}, {"$set": to_update})

    def get_all(self) -> list[Target]:
        """
        Get all targets

        Returns:
            list[Target]: _description_
        """
        rows = self.collection.find()
        targets = []
        for row in rows:
            row.update({"id": row["_id"]})
            targets.append(Target.model_validate(row))
        return targets

File: code/scanner/internal/repositories/target.go
package repositories

import (
	"context"
	"errors"
	"reflect"
	"time"

	"github.com/CSPF-Founder/shieldsup/scanner/enums"
	"github.com/CSPF-Founder/shieldsup/scanner/models"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
)

type TargetRepository struct {
	collection *mongo.Collection
}

func NewTargetRepository(db *mongo.Database) TargetRepository {
	return TargetRepository{
		collection: db.Collection("targets"),
	}
}

func (c *TargetRepository) FindByID(targetID string) (target models.Target, err error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	objectID, objectIDErr := primitive.ObjectIDFromHex(targetID)
	if objectIDErr != nil {
		return models.Target{}, objectIDErr
	}

	err = c.collection.FindOne(ctx, map[string]primitive.ObjectID{"_id": objectID}).Decode(&target)
	// Document found, return the target
	return target, err
}

func (c *TargetRepository) UpdateScanStatus(target *models.Target) (bool, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	if target == nil || reflect.TypeOf(target) != reflect.TypeOf(&models.Target{}) {
		return false, errors.New("invalid target object given") // Might need to change later
	}

	now := time.Now()
	toUpdate := bson.M{}

	switch target.ScanStatus {
	case enums.TargetStatusScanStarted:
		target.ScanStartedTime = now
		toUpdate["scan_started_time"] = now
	case enums.TargetStatusReportGenerated:
		target.ScanCompletedTime = now
		toUpdate["scan_completed_time"] = now
	}

	toUpdate["scan_status"] = target.ScanStatus

	filter := bson.M{"_id": target.ID}
	update := bson.M{"$set": toUpdate}
	_, updateError := c.collection.UpdateOne(ctx, filter, update)
	if updateError != nil {
		return false, updateError
	}
	return true, nil
}

func (c *TargetRepository) UpdateScanStatusByID(ctx context.Context, targetID primitive.ObjectID, scanStatus enums.TargetStatus) (bool, error) {
	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()

	now := time.Now()
	toUpdate := bson.M{}
	switch scanStatus {
	case enums.TargetStatusScanStarted:
		toUpdate["scan_started_time"] = now
	case enums.TargetStatusReportGenerated:
		toUpdate["scan_completed_time"] = now
	}

	filter := bson.M{"_id": targetID}
	update := bson.M{"$set": toUpdate}

	toUpdate["scan_status"] = scanStatus
	_, updateError := c.collection.UpdateOne(ctx, filter, update)
	if updateError != nil {
		return false, updateError
	}
	return true, nil
}

func (c *TargetRepository) MarkAsComplete(target *models.Target) (bool, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	now := time.Now()
	toUpdate := bson.M{}
	toUpdate["scan_completed_time"] = now
	toUpdate["scan_status"] = enums.TargetStatusReportGenerated
	toUpdate["overall_cvss_score"] = target.OverallCVSSScore

	if target.IsIPRange() {
		toUpdate["cvss_score_by_host"] = target.CVSSScoreByHost
	}
	filter := bson.M{"_id": target.ID}
	update := bson.M{"$set": toUpdate}

	_, updateError := c.collection.UpdateOne(ctx, filter, update)
	if updateError != nil {
		return false, updateError
	}
	return true, nil
}

func (c *TargetRepository) GetAll() (targets []models.Target, err error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	cursor, findErr := c.collection.Find(ctx, map[string]string{})
	if findErr != nil {
		return nil, findErr
	}
	defer cursor.Close(ctx)
	if err != nil {
		return nil, err
	}

	if err = cursor.All(ctx, &targets); err != nil {
		return nil, err
	}

	return targets, nil
}
